polkadot/node/core/pvf/execute-worker/src/lib.rs

```diff
@@ -343,28 +425,70 @@ fn handle_child_process(
 	send_child_response(&mut pipe_write, response);
 }
 
-/// Waits for child process to finish and handle child response from pipe.
+/// Returns stack size based on the number of threads.
+/// The stack size is represented by 2MiB * number_of_threads + native stack;
 ///
-/// # Arguments
+/// # Background
 ///
-/// - `pipe_read`: A `PipeReader` used to read data from the child process.
+/// Wasmtime powers the Substrate Executor. It compiles the wasm bytecode into native code.
+/// That native code does not create any stacks and just reuses the stack of the thread that
+/// wasmtime was invoked from.
 ///
-/// - `child`: The child pid.
+/// Also, we configure the executor to provide the deterministic stack and that requires
+/// supplying the amount of the native stack space that wasm is allowed to use. This is
+/// realized by supplying the limit into `wasmtime::Config::max_wasm_stack`.
 ///
-/// - `usage_before`: Resource usage statistics before executing the child process.
+/// There are quirks to that configuration knob:
 ///
-/// - `timeout`: The maximum allowed time for the child process to finish, in `Duration`.
+/// 1. It only limits the amount of stack space consumed by wasm but does not ensure nor check that
+///    the stack space is actually available.
+///
+///    That means, if the calling thread has 1 MiB of stack space left and the wasm code consumes
+///    more, then the wasmtime limit will **not** trigger. Instead, the wasm code will hit the
+///    guard page and the Rust stack overflow handler will be triggered. That leads to an
+///    **abort**.
+///
+/// 2. It cannot and does not limit the stack space consumed by Rust code.
+///
+///    Meaning that if the wasm code leaves no stack space for Rust code, then the Rust code
+///    will abort and that will abort the process as well.
+///
+/// Typically on Linux the main thread gets the stack size specified by the `ulimit` and
+/// typically it's configured to 8 MiB. Rust's spawned threads are 2 MiB. OTOH, the
+/// DEFAULT_NATIVE_STACK_MAX is set to 256 MiB. Not nearly enough.
+///
+/// Hence we need to increase it. The simplest way to fix that is to spawn an execute thread with
+/// the desired stack limit. We must also make sure the job process has enough stack for *all* its
+/// threads. This function can be used to get the stack size of either the execute thread or execute
+/// job process.
+fn get_max_stack_size(executor_params: &ExecutorParams, number_of_threads: u32) -> usize {
```

Comment:
1) See the comment on the preparation worker stack size; it shouldn't depend on the number of threads;
2) You could get the full semantics and extract the stack size here. There is no need to return the stack separately from `params_to_wasmtime_semantics()` as that's not an optimization: you process the full set of executor parameters to find out the stack size anyway.